DOC-1914, document the tree dialog component (#2656)
tinymce/tinymce-docs: modules/ROOT/pages/dialog-components.adoc

}
----

[[tree]]
=== tree

A *tree* is a composite component that presents an expandable category tree expandable down one level. It can also include a search and filter field.

> *Events:* Interacting with a *tree* component fires the `+onChange+` function in the dialog’s configuration.

Reviewer: ```suggestion
*onLeafAction:* Interacting with a *tree* component executes the `+onLeafAction+` callback set in the dialog configuration.
```
Not an event, executes a callback set using the `onLeafAction` option in the dialog spec.

            },
            items: [
                    {
      type: 'directory',
      id: 'dir',
      title: 'Directory',
      menu: {
        icon: 'image-options',
          type: 'menubutton',
          fetch: (success) => success([
            {
              type: 'menuitem',
              text: 'Menu item',
              onAction: () => {
                store.add('menuitem');
              }
            }
          ])
      },
      children: [
        {
          type: 'leaf',
          title: 'Leaf 3',
          id: '3',
        },
        {
          type: 'leaf',
          title: 'Leaf 4',
          id: '4',
          menu: {
            icon: 'image-options',
            type: 'menubutton',
            fetch: (success) => success([
              {
                type: 'menuitem',
                text: 'Menu item',
                onAction: () => {
                  store.add('menuitem');
                }
              }
            ])
          }
        },
      ]
    },
    {
      type: 'leaf',
      title: 'Leaf 5',
      id: '5',
    },
    {
      type: 'leaf',
      title: 'Leaf 6',
      id: '6',
    },
    ]

>     }]

Reviewer: This needs improved indentation to render correctly

Author: Re-formatted this example.

* a directory; and
* a leaf.

A directory item can contain other directory items and leaf items.

==== Activating items in a tree

Activating a directory item toggles the item between two states: collapsed and expanded.

Activating a leaf item triggers the `+onLeafAction+` callback.

Items are activated by clicking, pressing or by using keyboard-navigation.

==== Tree navigation and activation using a keyboard

* The **up–** or **down–arrow-key** highlights an item.
* If the highlighted item is a directory, the **left–** or **right–arrow-key** collapses or expands a directory item
* If the hightlighted item is a leaf, pressing **Return** triggers the `+onLeafAction+` callback.

*onLeafAction:* Interacting with a *tree* component executes the `+onLeafAction+` callback set in the dialog configuration.

[source,js]
----
tinymce.init({
  selector: "textarea",
  toolbar: "tree",
  setup: (ed) => {
    ed.ui.registry.addButton('tree', {
      text: 'Tree toolbar button',
      onAction: () => {
        ed.windowManager.open({
          title: 'Tree dialog',
          buttons: [{
	    type: 'cancel',
	    text: 'Cancel'
	  }],
          body : {

>             type: 'panel',
>             items: [{
>               type: 'tree',
>               onLeafAction: (id) => {
>                 console.log('clicked on item with id', id);
>               },
>               items: [
>                 {
>                   type: 'directory',
>                   id: 'dir',
>                   title: 'Directory',
>                   menu: {
>                     icon: 'image-options',
>                     type: 'menubutton',
>                     fetch: (success) => success([
>                       {
>                           type: 'menuitem',
>                           text: 'Menu item',
>                           onAction: () => {
>                             console.log('dir menu item');
>                           }
>                       }
>                     ])
>                   },
>                   children: [
>                     {
>                       type: 'leaf',
>                       title: 'Leaf 3',
>                       id: '3'
>                     },
>                     {
>                       type: 'leaf',
>                       title: 'Leaf 4',
>                       id: '4',
>                       menu: {
>                         icon: 'image-options',
>                         type: 'menubutton',
>                         fetch: (success) => success([
>                           {
>                             type: 'menuitem',
>                             text: 'Menu item',
>                             onAction: () => {
>                               console.log('leaf menu item'); // for demo purposes
>                             }
>                           }
>                         ])
>                       }
>                     },
>                   ]
>                 },
>                 {
>                   type: 'leaf',
>                   title: 'Leaf 5',
>                   id: '5'
>                 },
>                 {
>                   type: 'leaf',
>                   title: 'Leaf 6',
>                   id: '6'
>                 },

Reviewer: ```suggestion
tinymce.init({
  selector: "textarea",
  toolbar: "tree",
  setup: (ed) => {
    ed.ui.registry.addButton('tree', {
      text: 'Tree toolbar button',
      onAction: () => {
        ed.windowManager.open({
          title: 'Tree dialog',
          buttons: [{
	    type: 'cancel',
	    text: 'Cancel'
	  }],
          body : {
            type: 'panel',
            items: [{
              type: 'tree',
              onLeafAction: (id) => {
                console.log('clicked on item with id', id);
              },
              items: [
                {
                  type: 'directory',
                  id: 'dir',
                  title: 'Directory',
                  menu: {
                    icon: 'image-options',
                    type: 'menubutton',
                    fetch: (success) => success([
                      'bold',
                      'italic',
                      'underline',
                      {
                          type: 'menuitem',
                          text: 'Menu item',
                          onAction: () => {
                            console.log('dir menu item');
                          }
                      }
                    ])
                  },
                  children: [
                    {
                      type: 'leaf',
                      title: 'Leaf 3',
                      id: '3'
                    },
                    {
                      type: 'leaf',
                      title: 'Leaf 4',
                      id: '4',
                      menu: {
                        icon: 'image-options',
                        type: 'menubutton',
                        fetch: (success) => success([
                          'blocks',
                          'fontfamily',
                          'fontsize',
                          {
                            type: 'menuitem',
                            text: 'Menu item',
                            onAction: () => {
                              console.log('leaf menu item'); // for demo purposes
                            }
                          }
                        ])
                      }
                    },
                  ]
                },
                {
                  type: 'leaf',
                  title: 'Leaf 5',
                  id: '5'
                },
                {
                  type: 'leaf',
                  title: 'Leaf 6',
                  id: '6'
                },
```
You could still combine their usage into one example, since the `success` function takes an array of menu items, all of these could be included in the same arrays

}
----

[[tree]]
=== tree

A *tree* is a composite component that presents an expandable category tree expandable down one level. It can also include a search and filter field.

*onLeafAction:* Interacting with a *tree* component executes the `+onLeafAction+` callback set in the dialog configuration.

[source,js]
----

> tinymce.init({
>   selector: "textarea",
>   toolbar: "tree",
>     setup: (ed) => {
>       ed.ui.registry.addButton('tree', {
>       text: 'Tree toolbar button',
>       onAction: () => {
>       ed.windowManager.open({
>       title: 'Tree dialog',
>       buttons: [{type: 'cancel', text: 'Cancel',}],
>       body : {
>         type: 'panel',
>         items: [{
>           type: 'tree',
>           onLeafAction: (id) => {
>             console.log('clicked on item with id', id);
>           },
>           items: [{type: 'directory', id: 'dir', title: 'Directory',
>       menu: {
>         icon: 'image-options',
>           type: 'menubutton',
>           fetch: (success) => success([
>             {type: 'menuitem', text: 'Menu item', onAction: () => {store.add('menuitem');}}
>           ])
>       },
>       children: [
>         {type: 'leaf', title: 'Leaf 3', id: '3',},
>         {type: 'leaf', title: 'Leaf 4', id: '4',
>           menu: { icon: 'image-options', type: 'menubutton', fetch: (success) => success([
>             {type: 'menuitem', text: 'Menu item', onAction: () => {store.add('menuitem');}
>             }
>           ])
>         }
>       },
>     ]
>   },
>   {type: 'leaf', title: 'Leaf 5', id: '5',},
>   {type: 'leaf', title: 'Leaf 6',id: '6',},
>   ]
>   }]
>   }
>   })
>   }
>   })
>   }
> });

Reviewer: ```suggestion
tinymce.init({
  selector: "textarea",
  toolbar: "tree",
  setup: (ed) => {
    ed.ui.registry.addButton('tree', {
      text: 'Tree toolbar button',
      onAction: () => {
        ed.windowManager.open({
          title: 'Tree dialog',
          buttons: [{type: 'cancel', text: 'Cancel',}],
          body : {
            type: 'panel',
            items: [{
              type: 'tree',
              onLeafAction: (id) => {
                console.log('clicked on item with id', id);
              },
              items: [
                {
                  type: 'directory',
                  id: 'dir',
                  title: 'Directory',
                  menu: {
                    icon: 'image-options',
                    type: 'menubutton',
                    fetch: (success) => success([
                      {
                        type: 'menuitem',
                        text: 'Menu item',
                        onAction: () => {
                          store.add('menuitem');
                        }
                      }
                    ])
                  },
                  children: [
                    {
                      type: 'leaf',
                      title: 'Leaf 3',
                      id: '3'
                    },
                    {
                      type: 'leaf',
                      title: 'Leaf 4',
                      id: '4',
                      menu: {
                        icon: 'image-options',
                        type: 'menubutton',
                        fetch: (success) => success([
                          {
                            type: 'menuitem',
                            text: 'Menu item',
                            onAction: () => {
                              store.add('menuitem');
                            }
                          }
                        ])
                      }
                    },
                  ]
                },
                {
                  type: 'leaf',
                  title: 'Leaf 5',
                  id: '5'
                },
                {
                  type: 'leaf',
                  title: 'Leaf 6',
                  id: '6'
                },
              ]
            }]
          }
        });
      }
    });
  },
});
```
I also can't find where `store` is defined, have you tested this configuration to make sure that it works? Cos this looks like it'll throw an error when it tries to find `store`.

Author: > have you tested this configuration to make sure that it works?

[Yes](https://fiddle.tiny.cloud/U3iaab).

> Cos this looks like it'll throw an error when it tries to find store.

It throws some violations on loading, but I can’t get it throw an error:
<img width="1126" alt="console for fiddle tiny cloud   U3iaab" src="https://user-images.githubusercontent.com/99639563/229407938-16bb3fc3-56ea-4644-9699-7e1f9bba71ff.png">

Reviewer: Thanks for the fiddle. When you unfold down to leaf 4, then click on the three dots, then 'Menu item', the error is thrown. 
![image](https://user-images.githubusercontent.com/89239746/229408290-959860b2-8713-4ede-a2e7-e4b1d6c9de60.png)

Reviewer: The `store` variable causes an error and should be removed. It can be replaced by a globally declared object, if it is necessary to include this functionality.

Reviewer 2: I agree, that was my mistake I didn't notice that I left the store variable there. `store.add` calls can be replaced here with `console.log` calls for example to demo the usage.

Author: > I agree, that was my mistake I didn't notice that I left the store variable there. store.add calls can be replaced here with console.log calls for example to demo the usage.

I’ve edited the demo javascript accordingly.

}
----

[[tree]]
=== tree

> A *tree* is a composite component that presents an expandable category tree expandable down one level. It can also include a search and filter field.

Reviewer 2: This is not accurate. This description fits the advtemplate plugin.
The tree component does NOT include a search and filter field.
Also the tree component does NOT present expandable category, and it is not expandable down ONLY one level.

A more accurate description in my opinion would be:
A tree is a component that can be used to represent a tree hierarchy such as folders and files for example. 

A tree can have two types of tree items, a directory tree item and a leaf tree item.
A directory tree item can contain leaf items or other directory items.
Clicking on a directory toggles its expanding/collapsing state.
Clicking on a leaf item triggers the onLeafAction callback.

I apologize if I didn't make this wasn't clear before.

Author: This introduction has been completely re-written.

*onLeafAction:* Interacting with a *tree* component executes the `+onLeafAction+` callback set in the dialog configuration.

[source,js]
----
tinymce.init({
  selector: "textarea",
  toolbar: "tree",
  setup: (ed) => {
    ed.ui.registry.addButton('tree', {
      text: 'Tree toolbar button',
      onAction: () => {
        ed.windowManager.open({

>           title: 'Tree dialog',
>           buttons: [{type: 'cancel', text: 'Cancel',}],
>           body : {

Reviewer: ```suggestion
          title: 'Tree dialog',
          buttons: [{
	    type: 'cancel',
	    text: 'Cancel'
	  }],
          body : {
```
Missed these when I was writing.

            type: 'panel',
            items: [{
              type: 'tree',
              onLeafAction: (id) => {
                console.log('clicked on item with id', id);
              },
              items: [
                {
                  type: 'directory',
                  id: 'dir',
                  title: 'Directory',
                  menu: {
                    icon: 'image-options',
                    type: 'menubutton',

>                     fetch: (success) => success([
>                       {
>                         type: 'menuitem',
>                         text: 'Menu item',
>                         onAction: () => {
>                           console.log; // for demo purposes
>                         }
>                       }
>                     ])

Reviewer: ```suggestion
                    fetch: (success) => success([
		     'bold', 'italic', 'underline'
                    ])
```
These are menu items, so some that already exist can be called in this fetch function.

                  },
                  children: [
                    {
                      type: 'leaf',
                      title: 'Leaf 3',
                      id: '3'
                    },
                    {
                      type: 'leaf',
                      title: 'Leaf 4',
                      id: '4',
                      menu: {
                        icon: 'image-options',

>                         type: 'menubutton',
>                         fetch: (success) => success([
>                           {
>                             type: 'menuitem',
>                             text: 'Menu item',
>                             onAction: () => {
>                               console.log; // for demo purposes
>                             }
>                           }
>                         ])

Reviewer: ```suggestion
                        type: 'menubutton',
                        fetch: (success) => success([
			  'blocks', 'fontfamily', 'fontsize'
                        ])
```
Much simpler explanation than configuring custom menu items

Reviewer 2: @MitchCTiny I disagree. The point of having custom menu items there is to have a custom onAction that logs something to the console that shows that it works. One can not see the effect of these predefined menu items until they close the open modal. So to test things they would have to open the modal, click on the menu item then close it again to see if what they did had any effect. This is how I see it but I may be wrong 🤷‍♂️.

Author: I’ve added a second example. The first presents the tree as per @hamza0867’s original.

The second shows the tree making TinyMCE menu items available in the dialog.

You could, perhaps, accuse me of attempting to split the difference, but I think there’s a specific utility here — making menu items available in a dialog — that isn’t really about presenting a tree structure as such.

Reviewer: I'm happy with the use of both, including predefined menu items would show users the type of items shown here, while custom menu items show how it can be configured to their needs.
